jpeg/jpeg.go

```diff
@@ -0,0 +1,91 @@
+// jpeg handles scrubbing jpeg files of metadata
+package jpeg
+
+import (
+	"bytes"
+	"encoding/binary"
+	"io"
+	"io/ioutil"
+
+	jpegimagestructure "github.com/dsoprea/go-jpeg-image-structure"
+)
+
+var headerBytes []byte = []byte{0xff, 0xd8}
+
+type JpegScrubber struct {
+	reader      io.Reader
+	segmentData io.LimitedReader
+	wroteHeader bool
+}
+
+func NewJpegScrubber(reader io.Reader) *JpegScrubber {
+	return &JpegScrubber{
+		reader: reader,
+	}
+}
+
+func isMetadataType(segmentMarker []byte) bool {
+	// TODO: consider other APP types
+	return segmentMarker[0] == 0xff &&
+		(segmentMarker[1] == jpegimagestructure.MARKER_APP1 || segmentMarker[1] == jpegimagestructure.MARKER_APP0)
+}
+
+func (self *JpegScrubber) Read(buf []byte) (n int, err error) {
+	var n2 int
+
+	if !self.wroteHeader {
+		n, err = io.ReadFull(self.reader, buf[:len(headerBytes)])
+		if err != nil {
+			return
+		}
+		self.wroteHeader = true
+	}
+
+	n2, err = self.segmentData.Read(buf[n:])
+	n += n2
+	if err != io.EOF {
+		return
+	}
+
+	segmentHeaderBuffer := new(bytes.Buffer)
+	teedReader := io.TeeReader(self.reader, segmentHeaderBuffer)
+
+	reset := func() {
+		self.reader = io.MultiReader(segmentHeaderBuffer, self.reader)
+	}
+
+	segmentMarker := make([]byte, 2)
+	if _, err = io.ReadFull(teedReader, segmentMarker); err != nil {
+		reset()
```

Comment:
Why is it necessary to reset the reader here?  This reflects some kind of corruption in the file structure, right?  If `self.reader` is not returning well-formed JPEG data, it seems okay to me to just return an error.

Reply:
This goes for all of the reset logic by the way.  I'm not sure it helps the caller out at all and it adds complexity here.

Reply:
> Why is it necessary to reset the reader here? This reflects some kind of corruption in the file structure, right? If self.reader is not returning well-formed JPEG data, it seems okay to me to just return an error.

I think this would generally represent either corruption in the file structure OR an error reading from the upstream reader.

I'm still learning some go practices, but the thing I was worried about was returning the correct `n` bytes read value vis a vis the amount that the reader has been advanced. Is it "mandatory" to discard a reader after an error has been returned from it or is there some expectation of recovery? If so does `io.ReadFull` or `binary.Read` advance the reader? Am I over-thinking this???

Reply:
So I don't think the doc on `io.Reader` is crystal clear on this, but I believe the `n` value returned should be the number of bytes read _into the input byte slice_.  That is, if you read 20 bytes off the underlying reader and discard 5, you should return n = 15.

For example, think about a `tls.Conn`, which [reads off an underlying transport connection](https://golang.org/pkg/crypto/tls/#Client).  When I ask to read from the `tls.Conn`, I only want to know how many bytes were read into my byte slice.  I don't care that TCP flags were read, I just want the TLS payload.

I don't believe there is any expectation of recovery.  IMO, there's a bit of a contract in the API: you give me a JPEG reader, I'll scrub the metadata for you.  If a caller gives you a reader to a corrupt JPEG structure, the contract is violated and the best I can do for you is tell you so.  Same goes for other errors in the underlying reader.  I _think_ this is generally the case in Go API, but this is also kind of my opinion.  I think you're free to define recovery guarantees and strategies in your API.  However, I would shy away from that unless there was a strong and common use case you're trying to support.  

To add to that last point a little - since the caller gets to supply the underlying reader, they can always supply something that's going to buffer what's read and recover on error as they deem appropriate.  So callers aren't left high-and-dry if they want this kind of thing.

Reply:
Ok yeah I think that makes sense!